src/java.base/share/classes/java/io/BufferedInputStream.java

```diff
@@ -166,13 +170,17 @@ private InputStream getInIfOpen() throws IOException {
     }
 
     /**
-     * Check to make sure that buffer has not been nulled out due to
-     * close; if not return it;
+     * Return the buffer if already allocated, or allocate one, or throw {@link IOException} if this stream is closed.
+     * If a caller simply wants to ensure whether this BufferedInputStream is open
+     * they should call {@link #getInIfOpen()} instead.
      */
     private byte[] getBufIfOpen() throws IOException {
         byte[] buffer = buf;
         if (buffer == null)
             throw new IOException("Stream closed");
+        if (buffer == EMPTY) {
+            buf = buffer = new byte[size];
+        }
```

Comment:
You should probably use compareAndSet here too like in fill() - to handle the case of asynchoronous close.
This patch makes me a bit nervous. I would advise to get feedback from @bplb and @AlanBateman .

Reply:
You are probably following #13121, but for everybody else's reference here is Alan's comment on that PR:

https://github.com/openjdk/jdk/pull/13121#issuecomment-1481127986

Reply:
Yes, I'm nervous about changing BIS too. Most of the initial revisions of the patch introduced compatibility issues so it had to be restricted to the no sub-class case. Here, it needs to CAS buf as otherwise it will break async close. 

I think the starting point for this issue is to show examples where it avoids needing to create the byte[] on first usage. The first use of read, readAllBytes, readNBytes, .. will create the byte[] but Sergey may have spotted some usages where it does help.

Reply:
Some implementations of `getInputStream()` return `BufferedInputStream` which is later wrapped into `BufferedInputStream` on call site, e.g. `FileURLConnection`. With this change we can avoid allocation of byte[] in internal BIS.